### PR TITLE
feat(dmarc-reporter): add dmarc reporter

### LINKED ISSUE #230: Feature Request: DMARC report analyzer
Hey, @jogold, I just stumbled upon this repo. There are a bunch of really cool and helpful things here. Thanks for making them available!

One construct I've been working on is an automated way to parse DMARC report emails via a lambda + SES and notify a Slack channel of any failures. Unfortunately, I got sidetracked and still need to finish it.

I see an `EmailReceiver` and a bunch of Slack-related constructs here, so perhaps you would also be interested in this.

I'll see if I can finish mine and contribute it here. Is this something that would be interesting to merge?

### PR DESCRIPTION
**Pull request by andreialecu**
This is just a draft for now.

Closes #230

### CODE REVIEW

File: src/dmarc/index.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+export interface DmarcReporterProps {
+  /**
+   * The Route 53 hosted zone to create the DMARC record in.
+   */
+  hostedZone: route53.IHostedZone;
+  /**
+   * The email address to send DMARC reports to.
+   * This email address must be verified in SES.
+   */
+  emailAddress: string;
```
Comment (jogold) [inline src/dmarc/index.ts:14]: Can we default to `dmarc-reports` @ the domain name of the provided hosted zone?

File: src/dmarc/index.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  additionalEmailAddresses?: string[];
+
+  /**
+   * The DMARC policy to apply to messages that fail DMARC compliance.
+   * This can be one of the following values:
+   * - none: Do not apply any special handling to messages that fail DMARC compliance.
+   * - quarantine: Quarantine messages that fail DMARC compliance.
+   * - reject: Reject messages that fail DMARC compliance.
+   */
+  dmarcPolicy: 'none' | 'quarantine' | 'reject';
```
Comment (jogold) [inline src/dmarc/index.ts:28]: Let's create an enum.

```ts
export enum DmarcPolicy {
  NONE = 'none',
  ...
}
```

can we have default value?

File: src/dmarc/index.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * The DMARC policy to apply to messages that fail DMARC compliance for subdomains.
+   * This can be one of the following values:
+   * - none: Do not apply any special handling to messages that fail DMARC compliance.
+   * - quarantine: Quarantine messages that fail DMARC compliance.
+   * - reject: Reject messages that fail DMARC compliance.
+   * @default inherited from dmarcPolicy
+   */
+  dmarcSubdomainPolicy?: 'none' | 'quarantine' | 'reject';
```
Comment (jogold) [inline src/dmarc/index.ts:38]: enum

File: src/dmarc/index.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+   */
+  dmarcPercentage?: number;
+
+  /**
+   * The alignment mode to use for DKIM signatures.
+   * This can be one of the following values:
+   * - relaxed: Use relaxed alignment mode.
+   * - strict: Use strict alignment mode.
+   */
+  dmarcDkimAlignment?: 'relaxed' | 'strict';
```
Comment (jogold) [inline src/dmarc/index.ts:53]: enum

File: src/dmarc/index.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * The alignment mode to use for SPF signatures.
+   * This can be one of the following values:
+   * - relaxed: Use relaxed alignment mode.
+   * - strict: Use strict alignment mode.
+   */
+  dmarcSpfAlignment?: 'relaxed' | 'strict';
```
Comment (jogold) [inline src/dmarc/index.ts:61]: enum

File: src/dmarc/index.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+  /**
+   * The properties to pass to the EmailReceiver construct.
+   * The `recipients` property will be ignored.
+   */
+  emailReceiverProps: Omit<EmailReceiverProps, 'recipients'>;
```
Comment (jogold) [inline src/dmarc/index.ts:67]: Unfortunately the `Omit` will not work with JSII (used to provide the constructs in all the languages supported by the CDK), you will have to expose `receiptRuleSet`, `function` and `afterRule` in this interface.

**Review by jogold — COMMENTED** [review]
Nice! See my comments and thank you already for the contribution!

**Comment by andreialecu** [comment]
> Nice! See my comments and thank you already for the contribution!

Updated. If you'd like to merge it like this I could set it as Ready for review.

I was thinking about doing a full implementation using https://github.com/andreialecu/dmarc-report-parser but that could be left to the user I guess.

**Comment by andreialecu** [comment]
Added a snapshot test and updated docs as well. Should be good to go now. Up to you how you want to move forward with it. Let me know if you need anything else!

I'll probably just add something to the readme in https://github.com/andreialecu/dmarc-report-parser about how it can tie into this construct (and maybe update the docs for it here to link back there). But otherwise, I guess this is good to go like this.

**Review by jogold — APPROVED** [review]
Great!

**Pull request merged**